Refactor Actor::CalculateSpeed (WIP)

anewusername opened the pull request:
## Description

I took an initial pass at #901 .

At first glance most of the GUIScript.cpp calls (namely to `Inventory::RemoveItem` and `Inventory::AddSlotItem`) don't recalculate the weight, so there's not much simplification there. I will try to take a look at their other calling points to check if they could do so, unless someone recommends against it.


## Checklist

- [ ] Commit messages are descriptive and explain the rationale for changes
- [x] I used the same coding style as the surrounding code <!-- yet to be formally defined, see issue #161 -->
- [ ] I have tested the proposed changes
- [ ] I extended the documentation, if necessary
- [ ] The proposed change builds also on our build bots (check after submission)

lynxlynxlynx commented on gemrb/core/Inventory.cpp:
```diff
@@ -196,6 +196,8 @@ void Inventory::CalculateWeight() const
 	}
 	Changed = false;
+
+	Owner->SetStat(IE_ENCUMBRANCE, Weight, false);
```
This needs to be changed to SetBase, so we set the base stat. SetStat sets the modified one, which is wiped each game tick. This is probably why you couldn't just delete the calls from GUIScript (which is the way to go).

anewusername replied:
I added a commit changing it SetBase, but I'm not sure if it's necessary.
I think I was misunderstanding when CalculateWeight should be called (immediately before `GetWeight()`, gated by `Changed`, rather than on inventory update).
I also removed a few calls to CalculateWeight within other inventory functions to keep them consistent, since I couldn't find anywhere that the weight was accessed without first being recalculated.

lynxlynxlynx replied:
It feels a bit wrong — shouldn't be up to the inventory to keep track of its own state? Division of responsibilities and all that. Now it works only because the recalc is called each tick, so any changes are quickly picked up and there's at most a tick's delay.

lynxlynxlynx replied:
SetBase is needed even more than I thought too — if the inventory didn't change, nothing will set the stat that tick, so it will equal the base ... which is unset, so 0.

anewusername replied:
I forgot to add, maybe the right approach would be to call CalculateWeight from GetWeight?
That would ensure you don't get a stale weight, while encapsulating things better. Maybe `Weight` should also be private in that case.

lynxlynxlynx replied:
Weight should definitely remain private — for anyone that needs it, there's a getter already. CalculateWeight is a bit annoying, since it also does some flag cleanup, but it also rechecks everything (we don't update the weight atomically), which saves us having to deal with both the removed and new item if a swap occurred. I don't think it's worth touching it.

But I do think it's better to keep GetWeight const and make the effort to maintain an always correct Weight. `Changed` is there only for the weight handling, so it's easy to see where the updates should happen. Not sure why these two are marked mutable either.

anewusername replied:
I guess I see two paths for a clean interface:
- Use `Weight` to privately cache the value, invalidate using `Changed`, interface using `GetWeight`
- Remove `Changed`, replace Changed=True with calls to `CalculateWeight`. Less performant, but I doubt it'll make a difference.

lynxlynxlynx commented on gemrb/core/Scriptable/Actor.cpp:
```diff
@@ -5439,15 +5435,42 @@ int Actor::GetEncumbranceFactor(bool feedback) const
+	} else {
+		speed = 0;
+	}
+	if (core->HasFeature(GF_RESDATA_INI)) {
```
This block is completely separate even in Map. Since the function does two unrelated things at once, the proper thing is two add two more private functions and have CalculateSpeed basically be:
```
if (core->HasFeature(GF_RESDATA_INI)) {
  CalculateSpeedFromIni(feedback);
} else {
  CalculateSpeedFromRate(feedback);
}
```

lynxlynxlynx commented:
Thanks, a great start! Left some comments, so it can be improved.

lynxlynxlynx commented on gemrb/core/Scriptable/Actor.cpp:
```diff
@@ -5436,22 +5433,57 @@ int Actor::GetEncumbranceFactor(bool feedback) const
+		CalculateSpeedFromINI(feedback);
+	} else {
+		CalculateSpeedFromRate(feedback);
+	};
```
```suggestion
	}
```

lynxlynxlynx commented on gemrb/core/Scriptable/Actor.cpp:
```diff
@@ -5436,22 +5433,57 @@ int Actor::GetEncumbranceFactor(bool feedback) const
+	if (core->HasFeature(GF_ONE_BYTE_ANIMID)) {
+		animid = animid & 0xff;
+	}
+	if (animid < (ieDword)CharAnimations::GetAvatarsCount()) {
```
Since we're cleaning up, let's just make this an assert, since it should always be true. Will reduce the nesting by one level as well.

anewusername commented:
I refactored with the objective of keeping the weight/encumbrance updated on each slot update, and removed `Changed`.

I think it's reasonable to make `CalculateWeight` private and not const, since at this point it shouldn't need to be used externally.

lynxlynxlynx commented:
Looks good, thanks. Tried to break it, but didn't succeed, so even better. :)
